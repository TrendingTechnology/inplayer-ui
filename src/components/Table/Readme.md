Table is a component for displaying information in order user to look for patterns and easy to scan details. <br />
It takes columns and data props. Columns contain `title` - which is presented at the top od the columns and `key` for each column. Checkboxes should be added to each row if user wants to manipulate data. The data props contain object of each keys that are presented in the table cell according to the key.

```jsx static
const columns = [
  {
    title: '',
    key: 'assetType',
    render: value => <Icon name={value} className="icon" />,
  },
  {
    title: 'Name',
    key: 'name',
  },
  {
    title: 'ID',
    key: 'id',
  },
  {
    title: 'Date Created',
    key: 'dateCreated',
  },
  {
    title: 'Edit',
    key: 'edit',
    render: text => <a href="javascript:;">{text}</a>,
  },
];

const data = [
  {
    assetType: 'music',
    name: 'Andrej',
    id: 22,
    dateCreated: '2018 / 07 / 24',
  },
  {
    assetType: 'code',
    name: 'Petko',
    id: 28,
    dateCreated: '2018 / 07 / 21',
  },
];
<Table columns={columns} data={data} />;
```

```jsx
const columns = [
  {
    title: 'ID',
    key: 'id',
  },
  {
    title: '',
    key: 'icon',
    render: value => <Icon name={value} />,
  },
  {
    title: 'Name',
    key: 'name',
  },
  {
    title: 'Date Created',
    key: 'dateCreated',
  },
];

const data = [
  {
    id: 22,
    icon: 'music',
    name: "Won't You Be My Neighbour?",
    dateCreated: '2018/07/24',
  },
  {
    id: 28,
    icon: 'code',
    name: 'Soundtrack: No Country',
    dateCreated: '2018/07/21',
  },
];

<Table
  rowSelection={true}
  columns={columns}
  data={data}
  options={{
    rowSelection: {
      active: true,
      action: data => console.log(data),
    },
    rowActions: [
      { icon: 'edit', onClick: id => console.log(id) },
      { icon: 'cog', onClick: id => console.log(id) },
      { icon: 'trash', onClick: id => console.log(id) },
    ],
  }}
/>;
```
