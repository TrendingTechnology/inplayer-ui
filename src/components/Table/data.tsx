import React from 'react';
import { FaMusic } from 'react-icons/fa';

export const columns = [
  {
    title: 'ID',
    key: 'id',
  },
  {
    title: '',
    key: 'icon',
    render: ({ value }: any) => <FaMusic />,
  },
  {
    title: 'Name',
    key: 'name',
  },
  {
    title: 'Date Created',
    key: 'dateCreated',
  },
  {
    title: 'Director',
    key: 'director',
    render: ({ value, rowValues }: any) => `${value} - ${rowValues.dateCreated}`,
  },
];

export const data = [
  {
    id: 22,
    icon: 'music',
    name: "Won't You Be My Neighbour?",
    director: 'Morgan Neville',
    dateCreated: '2018/07/24',
  },
  {
    id: 28,
    icon: 'code',
    name: 'Soundtrack: No Country',
    director: 'Coen Brothers',
    dateCreated: '2018/07/21',
  },
];

export const totalPages = Math.ceil(data.length / 5);
