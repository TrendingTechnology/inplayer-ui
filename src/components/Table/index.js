// @flow
import React, { type Node } from 'react';
import styled from 'styled-components';
import colors from 'config/colors';
import { uiColors, fontSizes, fontWeights } from 'utils';
import Checkbox from 'components/Checkbox';
import Icon from 'elements/Icon';

const TableWrapper = styled.table`
  background: ${colors.white};
  width: 100%;
  text-align: left;
  border-collapse: collapse;
  color: ${uiColors('text.light')};
  box-sizing: border-box;
  border: 1px solid ${colors.gray};
  border-radius: 3px;
  position: relative;
  overflow-x: hidden;
  font-weight: ${fontWeights('light')};
`;

const TableHeadRow = styled.tr`
  border-bottom: 1px solid ${colors.gray};
  width: 100%;
`;

const TableRow = styled.tr`
  border-bottom: 1px solid ${colors.lightGray};
  width: 100%;

  &:last-child {
    border-bottom: none;
  }
`;

const TableCell = styled.td`
  font-size: ${fontSizes('medium')};
  padding: 7px 1%;
  vertical-align: middle;
  line-height: 30px;
  text-align: left;
  max-width: 450px;
`;

const TableHeaderCell = styled.th`
  padding: 8px 1%;
  font-weight: 300;
  font-size: ${fontSizes('small')};
  text-align: left;
  max-width: 450px;
`;

const ActionIcon = styled(Icon)`
  padding: 2px;
  cursor: pointer;
  transition: all 0.3s ease;

  &:hover {
    color: ${uiColors('primary.main')};
  }
`;

const TableCheckbox = styled(Checkbox)`
  label {
    top: -10px;
  }
`;

type Column = {
  title: string,
  key: string,
  render: ({ value: string }) => Node,
};

type Data = Object;

type RowAction = {
  icon: string,
  onClick: (id: number | string) => any,
};

type TableOptions = {
  rowSelection: {
    active: boolean,
    action: (selectedItems: Array<Data>) => any,
  },
  rowActions: Array<RowAction>,
};

type Props = {
  columns: Array<Column>,
  data: Array<Data>,
  className?: string,
  style?: Object,
  options?: TableOptions,
};

type State = {
  selected: {
    [number | string]: boolean,
  },
  selectedAll: boolean,
};

class Table extends React.Component<Props, State> {
  state = {
    selected: {},
    selectedAll: false,
  };

  toggleRow = (id: number) => () => {
    const { selected } = this.state;
    const isSelected = !selected[id];
    const newSelected = {
      ...selected,
      [id]: isSelected,
    };

    this.setState(({ selectedAll }) => ({
      selected: newSelected,
      selectedAll: !isSelected ? false : selectedAll,
    }));
  };

  toggleSelectAll = () => {
    const { selectedAll } = this.state;
    const { data } = this.props;
    const selected = data.reduce((acc, current) => ({ ...acc, [current.id]: !selectedAll }), {});

    this.setState(({ selectedAll: prevSelectedAll }) => ({
      selected,
      selectedAll: !prevSelectedAll,
    }));
  };

  generateRows = (data: Array<Data>) => {
    const { options } = this.props;
    const {
      rowSelection: { active },
      rowActions,
    } = options;
    let newData = [...data];

    if (active) {
      const { selected } = this.state;

      newData = data.map(dataCell => ({
        ...dataCell,
        check: (
          <TableCheckbox
            id={dataCell.id}
            checked={selected[dataCell.id]}
            onChange={this.toggleRow(dataCell.id)}
          />
        ),
      }));
    }

    if (rowActions) {
      newData = newData.map(dataCell => ({
        ...dataCell,
        actions: rowActions.map((action, index) => (
          <ActionIcon key={index} name={action.icon} onClick={() => action.onClick(dataCell.id)} />
        )),
      }));
    }

    return newData;
  };

  generateColumns = (columns: Array<Column>) => {
    const { selectedAll } = this.state;
    const { options } = this.props;
    const {
      rowSelection: { active },
      rowActions,
    } = options;

    let newColumns = [...columns];

    if (active) {
      const rowCheckbox = {
        title: (
          <TableCheckbox id="toggleAll" checked={selectedAll} onChange={this.toggleSelectAll} />
        ),
        key: 'check',
      };

      newColumns = [rowCheckbox, ...columns];
    }

    if (rowActions) {
      const actionsColumn = { title: 'Actions', key: 'actions' };
      newColumns = [...newColumns, actionsColumn];
    }

    return newColumns;
  };

  renderColumns = (columns: Array<Column>): Array<Node> =>
    this.generateColumns(columns).map((column, index) => (
      <TableHeaderCell key={index}>{column.title}</TableHeaderCell>
    ));

  renderData = (columns: Array<Column>, data: Array<Data>) => {
    const newColumns = this.generateColumns(columns);
    const newData = this.generateRows(data);

    return newData.map(row => (
      <TableRow key={row.id}>
        {newColumns.map((column, index) => (
          <TableCell key={index}>
            {column.render ? column.render({ value: row[column.key] }) : row[column.key]}
          </TableCell>
        ))}
      </TableRow>
    ));
  };

  render() {
    const { columns, data, className, style } = this.props;
    return (
      <TableWrapper className={className} style={style}>
        <thead>
          <TableHeadRow>{this.renderColumns(columns)}</TableHeadRow>
        </thead>
        <tbody>{this.renderData(columns, data)}</tbody>
      </TableWrapper>
    );
  }
}

Table.defaultProps = {
  className: '',
  style: {},
  options: {
    rowSelection: {
      active: false,
    },
    rowActions: [],
  },
};

export default Table;
