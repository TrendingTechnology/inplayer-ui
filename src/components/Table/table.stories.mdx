import { useState } from 'react';
import { Meta, Story, Preview, Props } from "@storybook/addon-docs/blocks";
import Table from '.'
import Pagination from '../Pagination'
import { columns, data } from './data.tsx'

<Meta title="Table" component={<Table/>} />


```jsx
import { Table } from '@inplayer-org/inplayer-ui';
const Page = () => (
  <Table />
)
```

<Preview withSource="open">
  <Story name="default">
    <>
    <Table
            columns={columns}
            data={data}
            options={{
                rowSelection: {
                    active: true,
                    action: (data) => console.log(data),
                },
                rowActions: [
                    { icon: "edit", onClick: (id) => console.log(id), render: () => (<h1>Hello</h1>) },
                    { icon: "cog", onClick: (id) => console.log(id), render: () => (<h1>Hello</h1>) },
                    { icon: "trash", onClick: (id) => console.log(id), render: () => (<h1>Hello</h1>) },
                ],
            }}
        />
        <Pagination
            onPageChange={console.log}
            totalItems={50}
            startPage={4}
            numberOfPagesDisplayed={10}
            itemsPerPage={12}
        />
      </>
  </Story>
</Preview>