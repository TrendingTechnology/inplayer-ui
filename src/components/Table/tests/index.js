import { generateSnapshots } from 'helpers';

import Table from '../index';

generateSnapshots(Table);
