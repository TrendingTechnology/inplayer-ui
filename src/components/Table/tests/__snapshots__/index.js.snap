// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table renders correctly 1`] = `
.c0 {
  background: #fff;
  width: 100%;
  text-align: left;
  border-collapse: collapse;
  color: #a0a3aa;
  box-sizing: border-box;
  border: 1px solid #d7dde5;
  border-radius: 3px;
  position: relative;
  overflow-x: hidden;
  font-weight: 300;
}

.c0 .entryname {
  color: #44494e;
}

.c0 .checkbox label {
  top: -10px;
}

.c1 {
  border-bottom: 1px solid #d7dde5;
  width: 100%;
}

.c3 {
  border-bottom: 1px solid #F6F8FB;
  width: 100%;
}

.c3:last-child {
  border-bottom: none;
}

.c3 .icon {
  margin-top: 4px;
  vertical-align: middle;
  color: #b9c1cb;
  line-height: 1;
  font-size: 1rem;
  -webkit-transition: ease all 500ms;
  transition: ease all 500ms;
  position: relative;
  top: 0;
}

.c3 .action:hover {
  color: #44494e;
}

.c4 {
  font-size: 1rem;
  padding: 7px 1%;
  vertical-align: middle;
  line-height: 30px;
  text-align: left;
  max-width: 450px;
}

.c2 {
  padding: 8px 1%;
  font-weight: 300;
  font-size: 0.875rem;
  text-align: left;
  max-width: 450px;
}

<table
  className="c0"
  style={Object {}}
>
  <thead>
    <tr
      className="c1"
    >
      <th
        className="c2"
      >
        Name
      </th>
    </tr>
  </thead>
  <tbody>
    <tr
      className="c3"
    >
      <td
        className="c4"
      >
        Eli
      </td>
    </tr>
  </tbody>
</table>
`;
